app/controllers/communicarts_controller.rb

```diff
@@ -19,69 +15,89 @@ def send_cart
   end
 
 
-  private
+  protected
 
   def validate_access
-    return if signed_in?
-
-    @token = ApiToken.find_by(access_token: params[:cch])
-    if !@token
-      raise AuthenticationError.new(msg: 'something went wrong with the token (nonexistent)')
-    elsif @token.expires_at && @token.expires_at < Time.now
-      raise AuthenticationError.new(msg: 'something went wrong with the token (expired)')
-    elsif @token.used?
-      raise AuthenticationError.new(msg: 'Something went wrong with the token. It has already been used.')
-    elsif @token.cart_id != params[:cart_id].to_i
-      raise AuthenticationError.new(msg: 'Something went wrong with the cart (wrong cart)')
-    else
-      sign_in(@token.user)
+    if !signed_in?
+      authorize(:api_token, :is_valid?, params)
+      # validated above
+      sign_in(ApiToken.find_by(access_token: params[:cch]).user)
+    end
+    # expire tokens regardless of how user logged in
+    tokens = ApiToken.joins(:approval).where(approvals: {
+      user_id: current_user, proposal_id: self.cart.proposal})
+    tokens.where(used_at: nil).update_all(used_at: Time.now)
+
+    authorize(self.cart.proposal, :can_approve_or_reject?)
+    # Imitate an authorization error
+    # will need to replace this when a new version of pundit arrives
+    if params[:version] && params[:version] != self.cart.proposal.version.to_s
+      ex = NotAuthorizedError.new("not allowd to update this proposal")
+      ex.query, ex.record = :is_up_to_date?, self.cart.proposal
+      raise ex
     end
   end
 
-  def user_id
-    if signed_in?
-      current_user.id
+  def token_validation_errors(query)
+    message = "Something went wrong with the token "
+    case query
+    when :exists?
+      message += "(nonexistent)"
+    when :is_not_expired?
+      message += "(expired)"
+    when :is_not_used?
+      message += "(already used)"
+    when :is_correct_cart?
+      message += "(wrong cart)"
+    end
+    flash[:error] = message
+    render 'authentication_error', status: 403
+  end
+
+  def user_access_errors(query, default_message)
+    case query
+    when :is_approver?
+      message = "Sorry, you're not an approver on #{self.cart.proposal.public_identifier}."
+    when :is_pending_approver?
+      message = "You have already logged a response for Cart #{self.cart.proposal.public_identifier}."
+    when :is_up_to_date?
+      message = "This request has recently changed. Please review the modified request before approving."
     else
-      @token.user_id
+      message = default_message
     end
+    flash[:error] = message
+    redirect_to cart_path(self.cart)
   end
 
-  def authentication_error(e)
-    flash[:error] = e.message
-    render 'authentication_error', status: 403
+  def auth_errors(exception)
+    if exception.record == :api_token
+      self.token_validation_errors(exception.query)
+    else
+      self.user_access_errors(exception.query, exception.message)
+    end
   end
```

Comment:
I agree; the error messages are tied very tightly to the policy that failed, for example. I'll tinker with this a bit.